Base/usr/share/man/man1/test.md

```diff
@@ -0,0 +1,94 @@
+
+## Synopsis
+
+```**sh
+$ test expression
+$ test
+$ [ expression ]
+$ [ ]
+```
+
+## Description
+
+`test` takes a given expression and sets the exit code according to its truthiness, 0 if true, 1 if false.
+An omitted expression defaults to false, and an unexpected error causes an exit code of 126.
+
+If `test` is invoked as `[`, a trailing `]` is _required_ after the expression.
+
+## Expressions
+
+The expression can take any of the following forms:
+
+### Grouping
+
+* `( <expression> )` value of expression
+
+### Boolean operations
+
+* `! <expression>` negation of expression
+* `<expression> -a <expression>` boolean conjunction of the values
+* `<expression> -o <expression>` boolean disjunction of the values
+
+### String comparison
+
+* `-n <string>` whether the string is non-empty
+* `-z <string>` whether the string is empty
+* `<string> = <string>` whether the two strings are equal
+* `<string> != <string>` whether the two strings not equal
+* `<string>` whether the string is non-empty, only allowed when it is the only expression specified
+
+### Integer comparison
+
+* `<integer> -eq <integer>` whether the two integers are equal
+* `<integer> -ne <integer>` whether the two integers are not equal
+* `<integer> -lt <integer>` whether the integer on the left is less than the integer on the right
+* `<integer> -gt <integer>` whether the integer on the left is greater than the integer on the right
+* `<integer> -le <integer>` whether the integer on the left is less than or equal to the integer on the right
+* `<integer> -ge <integer>` whether the integer on the left is greater than or equal to the integer on the right
+
+### File comparison
+
+* `<file> -ef <file>` whether the two files are the same (have the same inode and device numbers)
+* `<file> -nt <file>` whether the file on the left is newer than the file on the right (modification date is used)
+* `<file> -ot <file>` whether the file on the left is older than the file on the right (modification date is used)
+
+### File type checks
+
+* `-b <file>` whether the file is a block device
+* `-c <file>` whether the file is a character device
+* `-f <file>` whether the file is a regular file
+* `-d <file>` whether the file is a directory
+* `-p <file>` whether the file is a pipe
+* `-S <file>` whether the file is a socket
+* `-h <file>`, `-L <file>` whether the file is a symbolic link
+
+### File permission checks
+
+* `-r <file>` whether the curent user has read access to the file
+* `-w <file>` whether the curent user has write access to the file
+* `-x <file>` whether the curent user has execute access to the file
+* `-e <file>` whether the file exists
+
+
+Except for `-h/-L`, all file checks dereference symbolic links.
+
+
+## Options
+
+None.
+
+## Examples
+
+```sh
+# Conditionally do something based on the value of a variable
+$ test "$foo" = bar && echo foo is bar
+# Check some numbers
+$ test \( 10 -gt 20 \) -o \( ! 10 -ne 10 \) && echo "magic numbers!"
+```
```

Comment:
These examples are not great: our shell, for POSIX compliance, will need to have `[` and `test` as shell builtin.

If anuser were to try to run these examples in a POSIX compliant shell, these examples would not test the `/bin/test` utility, but rather the `test` shell builtin.

We should either not put `Examples` in this `man` page or replace `test` with `/bin/test` or `\test` (if we support it).

Maybe we don't need to care about this now, though.

Reply:
Or we could just slap a note on it that says "[ and test might be shell builtins"?
we don't support the `\foo` thing (yet?).

Reply:
> we don't support the \foo thing (yet?).

I don't know? Does the shell even have any bultins?

Reply:
Yeah, it has...
```
    __ENUMERATE_SHELL_BUILTIN(cd)      \
    __ENUMERATE_SHELL_BUILTIN(cdh)     \
    __ENUMERATE_SHELL_BUILTIN(pwd)     \
    __ENUMERATE_SHELL_BUILTIN(exit)    \
    __ENUMERATE_SHELL_BUILTIN(export)  \
    __ENUMERATE_SHELL_BUILTIN(unset)   \
    __ENUMERATE_SHELL_BUILTIN(history) \
    __ENUMERATE_SHELL_BUILTIN(umask)   \
    __ENUMERATE_SHELL_BUILTIN(dirs)    \
    __ENUMERATE_SHELL_BUILTIN(pushd)   \
    __ENUMERATE_SHELL_BUILTIN(popd)    \
    __ENUMERATE_SHELL_BUILTIN(time)    \
    __ENUMERATE_SHELL_BUILTIN(jobs)    \
    __ENUMERATE_SHELL_BUILTIN(disown)  \
    __ENUMERATE_SHELL_BUILTIN(fg)      \
    __ENUMERATE_SHELL_BUILTIN(bg)
```

and soon `alias`, once my shell PR gets merged.

Reply:
> These examples are not great: our shell, for POSIX compliance, will need to have [ and test as shell builtins.

It has been decided that our Shell has nothing to do with POSIX, and we come up with our own semantics that we believe to make sense. If the user needs a Bourne-style shell, they can use the GNU Bash port.